pkg/cloud/services/compute/instance.go

```diff
@@ -131,7 +132,28 @@ func (s *Service) CreateInstance(openStackCluster *infrav1.OpenStackCluster, mac
 		if err != nil {
 			return nil, err
 		}
-	} else {
+	}
+	if len(openStackMachine.Spec.Ports) > 0 {
+		for _, port := range openStackMachine.Spec.Ports {
```

Comment:
You can actually skip the `len(openStackMachine.Spec.Ports) > 0` check here, since your for loop will take that into consideration anyway. 

https://play.golang.org/p/Ry8T8bcDGmo